### PR TITLE
[Enhance] metainfo of dataset can be a generic dict-like Mapping

### PR DESCRIPTION
**Pull request by hiyyg**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

Allowing metainfo to be more generic, so that it is compatible with other dict-like config systems.

## Modification

Please briefly describe what modification is made in this PR.

Allowing metainfo to be `Mapping` (more generic), instead of just `dict`.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?

No

If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by zhouzaida** [comment]
Hi @hiyyg , thanks for your contribution. What additional mapping types do you want to support? Can you add the corresponding unit tests?

**Comment by hiyyg** [comment]
> Hi @hiyyg , thanks for your contribution. What additional mapping types do you want to support? Can you add the corresponding unit tests?

For example, when the metainfo is provided by a config dict created by omegaconf or other config systems, they are Mapping but not dict.

**Comment by hiyyg** [comment]
Where should the tests be added? Could you give me some guidances?

**Comment by zhouzaida** [comment]
> > Hi @hiyyg , thanks for your contribution. What additional mapping types do you want to support? Can you add the corresponding unit tests?
> 
> For example, when the metainfo is provided by a config dict created by omegaconf or other config systems, they are Mapping but not dict.

got it

**Comment by zhouzaida** [comment]
> Where should the tests be added? Could you give me some guidances?

Hi, you can add the unit tests in the following method. https://github.com/open-mmlab/mmengine/blob/daacb1878b5dc01023ff9390dca6c6744ad87e92/tests/test_dataset/test_base_dataset.py#L165.

File: mmengine/dataset/base_dataset.py [hunk 1]
```diff
@@ -155,8 +157,8 @@ class BaseDataset(Dataset):
 
     Args:
         ann_file (str, optional): Annotation file path. Defaults to ''.
-        metainfo (dict, optional): Meta information for dataset, such as class
-            information. Defaults to None.
+        metainfo (Union[Mapping, Config], optional): Meta information for
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:160]:
```suggestion
        metainfo (Mapping or Config, optional): Meta information for
```

File: mmengine/dataset/base_dataset.py [hunk 2]
```diff
@@ -213,7 +215,7 @@ class BaseDataset(Dataset):
 
     def __init__(self,
                  ann_file: Optional[str] = '',
-                 metainfo: Optional[dict] = None,
+                 metainfo: Optional[Union[Mapping, Config]] = None,
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:218]:
```suggestion
                 metainfo: Union[Mapping, Config, None] = None,
```

File: mmengine/dataset/base_dataset.py [hunk 3]
```diff
@@ -472,13 +474,13 @@ def load_data_list(self) -> List[dict]:
         return data_list
 
     @classmethod
-    def _load_metainfo(cls, metainfo: dict = None) -> dict:
+    def _load_metainfo(cls, metainfo: Union[Mapping, Config] = None) -> dict:
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:477]:
```suggestion
    def _load_metainfo(cls, metainfo: Union[Mapping, Config, None] = None) -> dict:
```

File: mmengine/dataset/base_dataset.py [hunk 4]
```diff
@@ -472,13 +474,13 @@ def load_data_list(self) -> List[dict]:
...
         """Collect meta information from the dictionary of meta.
 
         Args:
-            metainfo (dict): Meta information dict. If ``metainfo``
-                contains existed filename, it will be parsed by
-                ``list_from_file``.
+            metainfo (Union[Mapping, Config]): Meta information dict.
```
Comment (zhouzaida) [inline mmengine/dataset/base_dataset.py:481]:
```suggestion
            metainfo (Mapping or Config, optional): Meta information dict.
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**